src/compiler/type_checker.ml

```diff
@@ -187,79 +347,81 @@ let silence_linear_warn f =
   let len = String.length s in
   len > 0 && (s.[0] = '_' || s.[len-1] = '_')
 
-let check ~is_rule ~type_abbrevs ~kinds ~types:env ~unknown (t : ScopedTerm.t) ~(exp : TypeAssignment.t) =
+let checker ~type_abbrevs ~kinds ~types:env ~unknown :
+  (is_rule:bool -> ScopedTerm.t -> exp:TypeAssignment.t -> env_undeclared) * 
+  ((_, ScopedTerm.t) Ast.Chr.t -> env_undeclared) *
+  (ScopedTerm.t * Ast.Loc.t -> env_undeclared)
+=
+  let valid_mode = ref true in
+
   (* Format.eprintf "============================ checking %a\n" ScopedTerm.pretty t; *)
-  let sigma : (TypeAssignment.t * int * Loc.t) F.Map.t ref = ref F.Map.empty in
+  let sigma : sigma F.Map.t ref = ref F.Map.empty in
   let unknown_global = ref unknown in
   let fresh_name = let i = ref 0 in fun () -> incr i; F.from_string ("%dummy"^ string_of_int !i) in
   (* let set_fresh_id = let i = ref 0 in fun x -> incr i; x := Some !i in *)
 
-  let rec check (ctx : ret Scope.Map.t) ~loc ~tyctx x ety : spilled_phantoms =
-    (* Format.eprintf "@[<hov 2>checking %a : %a@]\n" ScopedTerm.pretty_ x TypeAssignment.pretty ety; *)
+  let global_type = global_type unknown_global in
+
+  let get_type ~loc ctx env (scope, c, _ as t) =
+    match scope with
+    | Scope.Global _ -> global_type env ~loc c
+    | Bound lang -> local_type ctx ~loc t
+  in
+
+  let rec check ~positive (ctx : ctx Scope.Map.t) ~loc ~tyctx x ety : spilled_phantoms =
+    (* Format.eprintf "@[<hov 2>checking %a : %a@]\n" ScopedTerm.pretty_ x TypeAssignment.pretty_mut_once ety; *)
     match x with
-    | Impl(b,t1,t2) -> check_impl ctx ~loc ~tyctx b t1 t2 ety
-    | Const(Global _ as gid,c) -> check_global ctx ~loc ~tyctx (gid,c) ety
-    | Const(Bound lang,c) -> check_local ctx ~loc ~tyctx (c,lang) ety
+    | Impl(b,t1,t2) -> check_impl ~positive ctx ~loc ~tyctx b t1 t2 ety
+    | Const(Global _,_,_ as c) -> check_global ctx ~loc ~tyctx c ety
+    | Const(Bound _,_,_ as c) -> check_local ctx ~loc ~tyctx c ety
     | CData c -> check_cdata ~loc ~tyctx kinds c ety
     | Spill(_,{contents = Phantom _}) -> assert false
-    | Spill(sp,info) -> check_spill ctx ~loc ~tyctx sp info ety
-    | App(Global _ as gid,c,x,xs) -> check_app ctx ~loc ~tyctx (c,gid) (global_type env ~loc c) (x::xs) ety 
-    | App(Bound lang as gid,c,x,xs) -> check_app ctx ~loc ~tyctx (c,gid) (local_type ctx ~loc (c,lang)) (x::xs) ety
-    | Lam(c,cty,t) -> check_lam ctx ~loc ~tyctx c cty t ety
+    | Spill(sp,info) -> check_spill ~positive ctx ~loc ~tyctx sp info ety
```

Comment:
I'he double checked the role of the positive boolean in `check`.
It is still unused. I propose the following fix:
```suggestion
    | Spill(sp,info) -> 
      if not positive then error ~loc "Spilling in negative position is forbidden";
      check_spill ~positive ctx ~loc ~tyctx sp info ety
```
The typechecker raises an error if a spilling is found in negative position.
I think this partially fixes: [#271](https://github.com/LPCIC/elpi/issues/271)